src/commands/insertAllFiles.ts

```diff
@@ -0,0 +1,70 @@
+import { ToastType } from "../types";
+
+const insertAllFiles = function(view, event, files, options) {
+  if (files.length === 0) return;
+
+  const {
+    dictionary,
+    uploadFile,
+    onFileUploadStart,
+    onFileUploadStop,
+    onShowToast,
+  } = options;
+
+  if (!uploadFile) {
+    console.warn("uploadFile callback must be defined to handle file uploads.");
+    return;
+  }
+
+  // okay, we have some dropped files and a handler – lets stop this
+  // event going any further up the stack
+  event.preventDefault();
+
+  // let the user know we're starting to process the files
+  if (onFileUploadStart) onFileUploadStart();
+
+  // we'll use this to track of how many files have succeeded or failed
+  let complete = 0;
+
+  const { state } = view;
+  const { from, to } = state.selection;
+
+  // the user might have dropped multiple files at once, we need to loop
+  for (const file of files) {
+    // start uploading the file file to the server. Using "then" syntax
+    // to allow all placeholders to be entered at once with the uploads
+    // happening in the background in parallel.
+    uploadFile(file)
```

Comment:
The delay between the user starting the upload and the files being inserted will result in strange and potentially broken scenarios – the document content can have changed substantially in that time.

With the image extension a placeholder widget is used to hold the position for this reason, but a better comparison might be the logic when creating a document from the link menu – a "placeholder" link is created and then updated when the document has been successfully created. A similar approach seems necessary here.

Reply:
let me know if I got this right:

You are suggesting that as soon as a user selects a file using the menu option you want me to insert a place holder to the document and save it. And then start the file upload process. Once the file is uploaded just replace the link.

is that what you are suggesting?

Reply:
Yes, here's the existing code:

https://github.com/outline/rich-markdown-editor/blob/7b031097489950e812f8b04e9df797b5030bcd68/src/components/LinkToolbar.tsx#L71-L88

A "placeholder" link is created with a temporary url while the document is being created. Once it's uploaded the href is swapped out. We could also target the temporary link in css to maybe fade it out or similar while the upload is happening.